src/commands/handlers/registration.js

```diff
@@ -366,19 +370,29 @@ const handlers = {
     },
 
     RPL_SASLLOGGEDIN: function(command, handler) {
-        if (handler.network.cap.negotiating === true) {
+        if (handler.network.cap.negotiating) {
             handler.connection.write('CAP END');
             handler.network.cap.negotiating = false;
         }
     },
 
-    ERR_SASLNOTAUTHORISED: function(command, handler) {
+    ERR_NICKLOCKED: function(command, handler) {
+        handleSaslFail(handler, 'nick_locked', command);
+    },
+
+    ERR_SASLFAIL: function(command, handler) {
```

Comment:
Why are you only adding the disconnect logic here?
There's other failures which now keep the connection open and at least according to the spec this error is not guaranteed if you already received a different error

Reply:
anytime handleSaslFail() is called it will disconnect if its called within cap negotiation (assuming sasl_disconnect_on_fail is true)